plugins/kotlin/fir/src/org/jetbrains/kotlin/idea/completion/contributors/FirDeclarationFromUnresolvedNameContributor.kt

```diff
@@ -0,0 +1,142 @@
+ * ```
+ * val unr<caret>
+ * println(unresolvedVar)
+ * ```
+ * This contributor would contribute `unresolvedVar` at caret position above.
+ */
+internal class FirDeclarationFromUnresolvedNameContributor(
+    basicContext: FirBasicCompletionContext,
+    priority: Int,
+) : FirCompletionContributorBase<FirRawPositionCompletionContext>(basicContext, priority) {
+    override fun KtAnalysisSession.complete(positionContext: FirRawPositionCompletionContext) {
+        val declaration = positionContext.position.getCurrentDeclarationAtCaret() ?: return
+        val referenceScope = referenceScope(declaration) ?: return
+
+        referenceScope.forEachDescendantOfType<KtNameReferenceExpression> { refExpr ->
+            ProgressManager.checkCanceled()
+            processReference(referenceScope, declaration, refExpr)
+        }
+    }
+
+    private fun KtAnalysisSession.processReference(
+        referenceScope: KtElement,
+        currentDeclarationInFakeFile: KtNamedDeclaration,
+        unresolvedRef: KtNameReferenceExpression
+    ) {
+        // In a block, references must be after the declaration. Therefore, we only offer completion for unresolved names that appear after
+        // the current cursor position.
+        if (referenceScope is KtBlockExpression && unresolvedRef.startOffset < parameters.offset) {
+            return
+        }
+        val name = unresolvedRef.getReferencedName()
+        if (!prefixMatcher.prefixMatches(name)) return
+        if (!shouldOfferCompletion(unresolvedRef, currentDeclarationInFakeFile)) return
+        if (unresolvedRef.reference?.resolve() == null) {
+            val lookupElement = LookupElementBuilder.create(name).suppressAutoInsertion()
+                .also { it.priority = ItemPriority.FROM_UNRESOLVED_NAME_SUGGESTION }
+            sink.addElement(lookupElement)
+        }
+    }
+
+    private fun KtAnalysisSession.shouldOfferCompletion(
+        unresolvedRef: KtNameReferenceExpression,
+        currentDeclarationInFakeFile: KtNamedDeclaration
+    ): Boolean {
+        val refExprParent = unresolvedRef.parent
+        val refExprIsCallRef = refExprParent is KtCallableReferenceExpression
+        val refExprIsCall = refExprParent is KtCallExpression
+        val refExprIsType = refExprParent is KtUserType
+        val receiver = if (refExprIsCall) refExprParent.getReceiverForSelector() else unresolvedRef.getReceiverForSelector()
+        return when (val symbol = currentDeclarationInFakeFile.getSymbol()) {
+            is KtCallableSymbol -> {
+                if (refExprIsType) return false
+                // If current declaration is a function, we only offer names of unresolved function calls. For property declarations, we
+                // always offer an unresolved usage because a property may be called if the object has an `invoke` method.
+                if (!refExprIsCallRef && !refExprIsCall && symbol is KtFunctionLikeSymbol) return false
```

Comment:
I think that those three booleans make the code hard to understand

It is not clear if they are mutually exclusive (they probably are), and how the different combinations are handled

Maybe it would be better to do it like this?

```kotlin
when {
  refExprParent is KtUserType -> { ... }
  refExprParent is KtCallExpression -> { ... }
}
```

This way, those three scenarios would be cleanly separated from each other

Also, it would be clear if we want to do something at all if none of the booleans is true

If such refactoring turns out too tedious, you can leave it like that

Reply:
Thanks for the suggestion! Let me use when and inline these booleans. Hopefully it's clearer now.